Perform image fetching using check/get sub-steps

vito opened the pull request:
## What does this PR accomplish?

Feature

This PR introduces a pattern for running sub-steps and uses it to fetch task `image_resource`s and resource type images using the `check` step and the `get` step, rather than using the image fetching logic under `atc/worker/` package.

## Changes proposed by this PR:

* Bump and re-enable `yarn benchmark`
* Refactor `(exec.RunStep).Run` to return `(bool, error)`, remove `.Succeeded`
  * This was technically not necessary but things felt nicer this way and I had been looking for a reason to refactor this. 😅
  * Having `.Succeeded` required steps to be stateful and has caused panics in the past (#5830).
* Refactor web UI `StepTree` to not use lenses and instead just store step state in a `Dict StepID State`.
  * My brain fried trying to reckon how lenses would work with dynamic sub-steps. It's a lot simpler this way.
  * ...and maybe more performant? Seems like this should have less consing. Haven't benchmarked tho.
* Add `(exec.RunState).Run` which executes an `atc.Plan`.
  * This was actually really easy to implement.
* Extend `(exec.RunState).NewLocalScope` to additionally create a local *artifact* scope.
  * This will also prevent artifact pollution with the `across` step!
* Change `(exec.TaskStep).Run` to fetch the configured `ImageResource` as sub-steps by running a `CheckPlan` followed by a `GetPlan` in a nested scope.
  * This adds two new build events, `ImageCheck` and `ImageGet`, which contain the build plans for the sub-steps.
  * `exec.CheckStep` stores the version as the result, which is then set in the `atc.GetPlan`.
  * The artifact fetched by the `exec.GetStep` is then passed to `atc/worker/` in the same way that the `image:` task step field works.
  * ...it's pretty neato!
* Add web UI support, which just renders the sub-steps in a new toggleable 'step initialization' section.
* Fix a few bugs with the build page: hotkey overlay + duration tooltip z-index.

Here's how it looks (expanded, w/ tooltip):

![image](https://user-images.githubusercontent.com/1880/96324821-a7b31b00-0ff1-11eb-8981-1341e72369a0.png)

## Contributor Checklist
<!--
Most of the PRs should have the following added to them,
this doesn't apply to all PRs, so it is helpful to tell us what you did.
-->
- [x] Followed [Code of conduct], [Contributing Guide] & avoided [Anti-patterns]
- [x] [Signed] all commits
- [x] Added tests (Unit and/or Integration)
- [ ] Updated [Documentation]
- [ ] Added release note (Optional)

[Code of Conduct]: https://github.com/concourse/concourse/blob/master/CODE_OF_CONDUCT.md
[Contributing Guide]: https://github.com/concourse/concourse/blob/master/CONTRIBUTING.md
[Anti-patterns]: https://github.com/concourse/concourse/wiki/Anti-Patterns
[Signed]: https://help.github.com/en/github/authenticating-to-github/signing-commits
[Documentation]: https://github.com/concourse/docs

## Reviewer Checklist
<!--
This section is intended for the reviewers only, to track review
progress.
-->
- [ ] Code reviewed
- [ ] Tests reviewed
- [ ] Documentation reviewed
- [ ] Release notes reviewed
- [ ] PR acceptance performed
- [ ] New config flags added? Ensure that they are added to the
  [BOSH](https://github.com/concourse/concourse-bosh-release) and
  [Helm](https://github.com/concourse/helm) packaging; otherwise, ignored for
  the [integration
  tests](https://github.com/concourse/ci/tree/master/tasks/scripts/check-distribution-env)
  (for example, if they are Garden configs that are not displayed in the
  `--help` text).

aoldershaw commented on web/elm/src/Build/StepTree/StepTree.elm:
```diff
@@ -774,6 +808,22 @@ viewStepWithBody model session depth { id, name, log, state, error, expanded, ve
+                            ]
+
+                    Nothing ->
+                        Html.text ""
```
Not sure what the best design is here, but IMO it's a bit hard to tell that the build output for the `task` step belong to the task step itself, rather than the `get: image` (when the `get: image` header is collapsed). Sure, build logs are indented, but it's pretty subtle:

<img width="801" alt="Screen Shot 2020-10-14 at 12 08 22 PM" src="https://user-images.githubusercontent.com/26583442/96017007-72eb6c00-0e17-11eb-987e-cccb1455ab59.png">

I wonder if you could just throw in something to separate them a bit more (could just be a `<hr>`, or more padding, or *something*) - not sure if @matthewpereira has any ideas.

p.s. I like nesting the headers under `task`, esp. with the sticky headers

vito replied:
Yeah good point. (btw, what happened with the task name in your screenshot?)

I've got a couple other things to fix up (added a TODO section above) so I'll see if I can think of something to make this a little more obvious in the meantime.

aoldershaw replied:
> btw, what happened with the task name in your screenshot?

I was initially wondering if something in this PR broke the task header, but I realized I just had `task: ""` in the pipeline config I used

aoldershaw replied:
Though interestingly it didn't trigger a config validation warning...but we probably should

aoldershaw reviewed (changes requested):
This is really cool to see in action! I'm excited about the opportunities this brings for the future. Also, nice refactor of the `StepTree`/foci implementation - it took me a while to wrap my head around it when I first saw it.

Just have a couple minor things

vito commented on atc/exec/across_step.go:
```diff
@@ -127,20 +127,8 @@ func (step AcrossStep) acrossStepLeafExecutor(state RunState, steps []ScopedStep
-		if !s.Succeeded() {
-			return false
-		}
+			return steps[i].Run(ctx, scope)
```
Haha, yeah when I saw these type signatures it felt like the stars coming into alignment. 😄

vito commented on atc/exec/task_step.go:
```diff
@@ -303,12 +309,68 @@ func (step *TaskStep) imageSpec(logger lager.Logger, repository *build.Repositor
+
+			delegate.ImageCheck(lagerctx.FromContext(ctx), checkPlan)
+
+			ok, err := state.Run(ctx, checkPlan)
```
Right?! :D It's even cooler in the later commits which change these two steps to run in a nested scope so they can't pollute the outer plan state. The `get` step fetches an artifact named `image` and the outer `task` step grabs it from the nested scope.

vito commented on atc/exec/build/repository.go:
```diff
@@ -20,13 +20,16 @@ type ArtifactName string
 type Repository struct {
 	repo  map[ArtifactName]runtime.Artifact
-	repoL sync.RWMutex
+	repoL *sync.RWMutex
```
I can undo this actually. I originally did it because my first implementation of `NewLocalScope` involved cloning the parent like this:

```go
clone := *repo
```

My editor threw a warning for this because it copied the mutex, so I changed the mutex to a pointer and reassigned it.

But then I ended up changing it to just `clone := NewRepository()` and assigning `.parent` after, so this change isn't needed anymore. I'll just undo it to prevent code churn.

aoldershaw replied:
Cool, I don't really have a preference, was mostly just wondering if you knew something I didn't 😁

aoldershaw commented on atc/exec/put_step.go:
```diff
@@ -220,15 +220,7 @@ func (step *PutStep) run(ctx context.Context, state RunState, delegate PutDelega
-
-// Succeeded returns true if the resource script exited successfully.
-func (step *PutStep) Succeeded() bool {
-	return step.succeeded
```
Took a look at the deepsource output, and sure enough, it pointed out the useful fact that we can remove `step.succeeded` from `PutStep`. Thanks, deepsource!

https://deepsource.io/gh/concourse/concourse/run/970748ed-d236-4715-bd3c-c5ea58a0bc52/go/SCC-U1000

vito commented:
For shits and giggles, I'm working on a followup to this that uses the same approach for `check`, `get`, and `put` step resource type images. Using it for all 4 container steps will let us completely remove the worker image fetching logic.

It's not a whole lot of work, but it would increase the scope of this PR a bit. I think I'll finish it up and push it anyway, but I'm happy to split it out if it makes reviewing easier. I'm a little motivated to do this sooner rather than later because I'd like to be able to move the image policy checking code into one place rather than duplicate it between the worker image fetcher and the new flow.

Fetching resource type images this way would also remove the need for creating builds to check resource types themselves, since they could be checked everywhere on-demand (while still respecting an interval - the `check` step already [knows to no-op](https://github.com/concourse/concourse/blob/1746cf68a49e35fde96f8b7ecde873b517abf2f4/atc/engine/builder/check_delegate.go#L118-L138)). Checking on-demand also removes the 'resource type has no version' error condition, which is kinda neat.

This deeply-nested-types pipeline looks pretty wild:

```yaml
---
resource_types:
- name: registry-image-2
  type: registry-image
  source: {repository: concourse/registry-image-resource}

- name: git
  type: registry-image-2
  source: {repository: concourse/git-resource}

resources:
- name: booklit
  type: git
  source: {uri: "https://github.com/vito/booklit"}

jobs:
- name: unit
  plan:
  - get: booklit
    trigger: true
  - task: test
    file: booklit/ci/test.yml
```

![image](https://user-images.githubusercontent.com/1880/96055939-62f17d80-0e53-11eb-9d23-9c1adbd55a78.png)

aoldershaw commented:
> I'm working on a followup to this that uses the same approach for `check`, `get`, and `put` step resource type images. Using it for all 4 container steps will let us completely remove the worker image fetching logic.

Sounds cool! Although, I will say the UX might be a bit confusing. In the screenshot you provided, it may not be obvious to new Concourse users why a `get` would have a nested `get`, which may also have a nested `get`, and so on. I wonder if there are UX niceties we could implement to make it clear that the `get`s and `check`s are implicit, since I can see people being confused thinking "I didn't put a `get` here".

Another potentially confusing aspect is that all the steps are all named `get: image`. What if we name them after the resource that we're actually getting instead. So, `get: booklit` > `get: git` > `get: registry-image-2`. For tasks, maybe we could have `get: image_resource` (and `check: image_resource`), so at least it matches the field name in the `config`.

> I'm happy to split it out if it makes reviewing easier

No need for that, happy to do it all in one go. Already gone through most of the changes, happy to keep reviewing

evanchaoli commented:
TBH, this GUI feels very confused. Users won't care about how an image is fetched at all, the sub-steps make the GUI messy and hard to understand. Do we have to show those sub-steps to users?

vito commented:
> TBH, this GUI feels very confused. Users won't care about how an image is fetched at all, the sub-steps make the GUI messy and hard to understand. Do we have to show those sub-steps to users?

Users will care when an image fails to check or get. The UI here is not final - I agree it's too in-your-face. We'll likely land on something much more subtle, maybe an icon tucked into the header that users only notice for the 1 in 1000 builds that it fails. But when it fails, I think it's important for there to be something to see.

vito commented:
The last batch of commits move the image fetching from the `task` step to `(exec.BuildStepDelegate).FetchImage` so it can be used by the `get`, `check`, and `put` steps. Next the image policy checking is added to the delegate, and then all the worker image fetching logic is removed, as nothing uses it anymore. (🎉)

There's more cleanup that could be done: I think we can remove the entire `atc.VersionedResourceTypes` type and stop performing resource type checks. I think we could also get rid of `worker.Fetcher`, which used to exist because of the image fetching code, but now is a kind of unnecessary indirection - its locking and caching behavior could be moved to the `get` step, where it'd be kind of symmetrical to the `check` step's locking and no-op behavior.

But I'm going to pause before doing that and try to get the UI to a better place so this can be in a mergeable state first. The current UI is really confusing. 😅 I think what I'll try to do is show a more useful name than `image` as you suggested @aoldershaw - the type name (e.g. `git`) for resource steps, or for a task, `image_resource` or `image` (not sure which - I like that `image_resource` is more explicit, but it is longer, so I'll have to see how it looks). The steps will be hidden by default, and the name will be a toggle in the header, something like `get: concourse [git]` where the `[git]` is clickable to reveal the image fetching steps, and *maybe* has some sort of status UI cue.

Another thing for me to remember is to add a configuration for the interval on which images should be checked. Right now it just checks all the time, but I could add a flag, maybe even with a higher default value than 1m as these things don't tend to change as often, and manually-triggered builds will forego the interval check anyway.

vito commented:
Alright, I think I got something that's not too shabby.

I added another icon (hoo boy) to the step header, to the left of the step status. I re-used the 'cogs' icon from the build preparation - my rationalization is that this is kind of 'step preparation', but we could always change the icon. I think this might be a good place to show not only image fetching but e.g. `get_var` steps in the future, when var sources are implemented by prototypes and may be run as sub-steps when evaluating `((vars))` syntax.

By default, the additional icon is the only visual difference. It can be clicked to reveal the initialization steps, which appear beneath the header with a slightly darker background (hooray 50 shades of grey) to differentiate the initialization steps from the outer step output (if it were to also be toggled open).

Along the way I noticed the step status tooltip (the one that shows initialization/run time) wasn't showing up properly, so I refactored it to use the (I'm guessing) newer tooltip flow.

Here's a bunch of screenshots!

Initialization closed, outer step output closed (default):

![image](https://user-images.githubusercontent.com/1880/96324275-fa3f0800-0fee-11eb-922d-0f9280b7d835.png)

Initialization closed, outer step output open:

![image](https://user-images.githubusercontent.com/1880/96324406-bbf61880-0fef-11eb-9cdb-2afc23f9faa5.png)

Initialization open:

![image](https://user-images.githubusercontent.com/1880/96324282-06c36080-0fef-11eb-87a9-09b172eda2ed.png)

Initialization closed, outer step output open:

![image](https://user-images.githubusercontent.com/1880/96324364-7cc7c780-0fef-11eb-8da0-c37dfb34e071.png)

Initialization open, inner step output:

![image](https://user-images.githubusercontent.com/1880/96324377-8f420100-0fef-11eb-8d7b-e1d136c552bc.png)


Tooltip:

![image](https://user-images.githubusercontent.com/1880/96324294-1773d680-0fef-11eb-8b7c-770f516a5348.png)

Intentionally confusing nested-resource-types example:

![image](https://user-images.githubusercontent.com/1880/96324328-4db15600-0fef-11eb-82ad-4403d2353bb6.png)

At this point the PR has gotten pretty huge so I'll pause before moving on to more refactoring. 😅

xtreme-sameer-vohra reviewed (commented):
wow thats a big one !!
the worker client side changes and the visualizing the steps in the UX is awesome

vito commented on atc/engine/builder/builder.go:
```diff
@@ -33,34 +31,32 @@ func NewStepBuilder(
 }
 
-type stepBuilder struct {
+type StepBuilder struct {
```
It doesn't, but tools like go lint/vet complain if you have a public function return a private type, and I personally find that pattern awkward too, so I just exported it along the way.

vito commented on atc/engine/builder/builder.go:
```diff
@@ -33,34 +31,32 @@ func NewStepBuilder(
 }
 
-func (builder *stepBuilder) BuildStep(logger lager.Logger, build db.Build) (exec.Step, error) {
-	if build == nil {
```
Yeah, I really don't know why this was there - it seemed like defensive programming. This is never called with a nil build.

vito commented on atc/engine/engine.go:
```diff
@@ -215,36 +215,38 @@ func (b *engineBuild) Run(ctx context.Context) {
 		}
-		b.finish(logger.Session("finish"), err, step.Succeeded())
+
+		b.finish(logger.Session("finish"), err, succeeded)
```
Is the idea that we don't want to risk setting `succeeded` to a value other than `false`? Technically that would make this part more airtight, but IMO it should be an invariant that when err != nil all other return values are zero-value. Given that assumption, I wouldn't expect to see a test for when a step returns `true, <some err>` so I wouldn't write the code to handle it either.

For that invariant to hold true I guess I would argue that all step tests should assert that the return value is false whenever an error is returned, rather than having tests in the API consumers cover an 'true, error' combination.

Does that make sense?

Code style wise it might make a bit more sense if I had done `var succeeded bool` instead of `succeeded := false` - the direct assignment here is just meant to set the value in the outer scope, which might be more obvious if it's not given an explicit value before. Kind of like this pattern:

```go
var succeeded bool
var err err
if runStep1 {
  succeeded, err = step1.Run()
} else {
  succeeded, err = step2.Run()
}
```

...but in this case `err` is already declared. Maybe I could give it its own name for clarity. 🤔 

```go
var succeeded bool
var stepErr error

done := make(chan struct{})
go func() {
  defer close(done)

  succeeded, stepErr = state.Run(...)
}()
```

xtreme-sameer-vohra replied:
Thanks. And looking at `b.finish`, it does handle the cases as well.

vito commented on atc/exec/build/repository.go:
```diff
@@ -68,3 +79,13 @@ func (repo *Repository) AsMap() map[ArtifactName]runtime.Artifact {
+	return child
+}
+
+func (repo *Repository) Parent() *Repository {
```
Yeah I'm iffy on it too - I did the same for `(exec.RunState).Parent`). It's a bit of a trade-off. Making it a tested API contract allows the tests for things which use `NewLocalScope()` to just call `.Parent()` to assert that they're nested. The alternative I see is to behaviorally test the artifact scoping mechanics in each place, but that seemed a bit too repetitive and brought the tests beyond their core competency (IMO `NewLocalScope` should be cheap to use and test).

So at this point it's an API contract covered in tests, and right now it's only there to make other tests a bit easier to write. It's not used in production, but it also shouldn't require any maintenance. It does make it easier to for someone to use it in the future, but it's also so trivial and that it would probably just be exported/added if it's not already there, so 🤷‍♂️ I'm OK with the trade-offs myself. Happy to chat more tho.

xtreme-sameer-vohra replied:
Hehe, you did add behavioural tests for `AsMap` and `ArtfactFor` in `repository_test.go` and it looks good :)
So, I feel like it should be safe to drop `Parent`. I agree it is trivial to add in but the action would have to be intentional. But overall, this is minor.

vito commented on atc/engine/builder/builder.go:
```diff
@@ -33,34 +31,32 @@ func NewStepBuilder(
-		return exec.IdentityStep{}, errors.New("must provide a build")
-	}
-
+func (builder *StepBuilder) BuildStepper(build db.Build) (exec.Stepper, error) {
```
Haha yeah these names got a little wack. I'll give it a think!

vito replied:
I just split this out into https://github.com/concourse/concourse/pull/6174

aoldershaw commented on web/elm/src/Build/Build.elm:
```diff
@@ -686,6 +696,19 @@ tooltip _ { hovered } =
+                { body =
+                    Html.div
+                        Styles.changedStepTooltip
+                        [ Html.text "initialization" ]
```
Question: is volume streaming considered part of initialization? And if so, might it be confusing to call the image fetching panel initialization? If the initialization takes 1 minute, but the `check` and `get` steps only take 10 seconds, I can see people being confused where the other 50 seconds wound up

vito replied:
Hmm yeah true. I was thinking we might add more initialization information there in the future, but I can change it to 'image fetching' or something so it's accurate until then.

It might be a good place to move the selected worker name, which was added in #5846 and is currently printed to the logs as a stopgap. I hacked it together for a sec but couldn't figure out a good UI quickly so I bailed on it for now.

aoldershaw replied:
I like the idea of adding more initialization info in the future - would be cool to add e.g. progress bars tracking artifact streaming. Before I understood what Concourse was doing a bit better, I was confused why a step could be spinning for a while with no logs being emitted, so revealing what's going on behind the scenes in the UI is a big plus

evanchaoli commented:
> > TBH, this GUI feels very confused. Users won't care about how an image is fetched at all, the sub-steps make the GUI messy and hard to understand. Do we have to show those sub-steps to users?
> 
> Speak for yourself! 😆 I will care when an image fails to check or get, or when a 2GB image takes 5 minutes to download and I want to see something more than a spinner on a task step so I know it's not just stuck. Or when I've misconfigured resource type, I'd like to see something better than "resource type has no version" so I can troubleshoot. With this change the root cause will be visible directly on the resource page output.
> 
> The UI here is not final - I agree it's too in-your-face. We'll likely land on something much more subtle, maybe an icon tucked into the header that users only notice for the 1 in 1000 builds that it fails. But when it fails, I think it's important for there to be something to see.

I might not state clearly. If you make a survey on how many users understand that task image is fetched via `registry-image`, I guess the number would be low. That's reason why I feel that showing `check` and `get` would make user confused. 

I'm not against your refactor, neither the idea of enhancing troubleshooting. I was just trying to avoid user confusing. Anyway, that's only my few cents, you may ignore.

aoldershaw commented on web/elm/src/Build/StepTree/StepTree.elm:
```diff
@@ -1257,41 +1234,76 @@ viewStepHeaderLabel headerType changed stepID =
+                        )
+
+                ( Nothing, Nothing, Nothing ) ->
+                    Html.text "pending"
```
Really minor, but for steps that'll never run (e.g. hooks that aren't run, steps after a failed step in a `do`, implicit `get` after failed `put`), the `pending` tooltip is a bit misleading:

<img width="502" alt="Screen Shot 2020-10-18 at 10 24 03 PM" src="https://user-images.githubusercontent.com/26583442/96395027-8426d680-1191-11eb-8e29-e3257e0a6b42.png">

aoldershaw replied:
Unrelated, and you may have noticed this, but the expanded icon doesn't look so great when the step header has a border:

<img width="490" alt="Screen Shot 2020-10-18 at 10 38 44 PM" src="https://user-images.githubusercontent.com/26583442/96395536-dddbd080-1192-11eb-857e-a42d1df68112.png">

vito replied:
Dang, yeah that's because I changed the border to a `box-shadow` because I noticed the headers became 1px bigger on all sides, bringing them out of alignment with other elements on the resource page. But now the shadow is below that element.

I think I'll change it back to borders, since I also noticed some weirdness on HiDPI displays - it seemed like the upper border was ~0.5px thicker than the lower one, and I'm guessing that's just more weirdness with box-shadow being used this way.

Will patch up the tooltip too. I'll have it show the step state if the timing info isn't available.

vito replied:
Fixed the border interaction...kind of:

![image](https://user-images.githubusercontent.com/1880/96473255-6abc7380-11ff-11eb-8bea-e353feec35f6.png)

The trade-off is that the background color isn't flush to the edges when the step has no border:

![image](https://user-images.githubusercontent.com/1880/96473356-8a539c00-11ff-11eb-9157-1fcdcf61e313.png)

...but I think I'm OK with this since it's at least consistent, and fixing it would involve CSS trickery that's dependent on the step's state. I don't want to get too lost in the weeds on this since we might adjust it more later anyway.

aoldershaw reviewed (approved):
LGTM, but I'll let Sameer finish up his review

xtreme-sameer-vohra commented:
Thanks, lgtm as well !!
I'll let others speak about the UX as my thoughts were basic 😺 
This PR is already huge and it moves the needle in the right direction. 
However, with regards to other refactoring ideas brought up, one pattern that I found a bit harder to reason about was `RunState` having the `Run` function on it. I wonder if that can be pulled out into another component 🤔  The last change of pulling the fetching logic into `build_step_delegate` already minimizes the surface area for this.

vito commented:
@xtreme-sameer-vohra `Run` felt most natural on `RunState` because it allows steps to easily run inner plans, which is inherently dependent on the state already. I also changed the engine to run the entire build (i.e. the 'root step') by calling `(exec.StepState).Run` so that it's consistently used everywhere.

I like this model because it allows for really really simple implementations of things like sequences. For example, we could implement `do:` as a type alias to `[]atc.Plan` rather than building up a chain of `exec.OnSuccessStep`s:

```go
type DoStep []atc.Plan

func (plans DoStep) Run(ctx context.Context, state RunState) (bool, error) {
  for _, plan := range plans {
    ok, err := state.Run(ctx, plan)
    if err != nil {
      return false, err
    }

    if !ok {
      return false, nil
    }
  }

  return true, nil
}
```

If we had to inject a `Runner` (or something) this wouldn't be able to be a type alias, since it'd have to hold the struct field for it. It's a tiny concern, but to me it's a good sign when you can leverage type aliases like this in a design that's so reliant on composition.

re: `(StepBuilder).BuildStep`, I have a commit ready that renames all the things, but now the entire `builder/` package name doesn't make sense. 😅 I'm thinking of just merging `engine/builder/` into `engine/`, but that's a lot more code churn so I think I'll do it in another PR since this one's already approved.

Here are the names I'm thinking:

| old | new |
|-|-|
| `builder.StepFactory` | `engine.CoreStepFactory` |
| `builder.StepBuilder` | `engine.StepperFactory` |

The `CoreStepFactory` re-uses the "core" term from the `atc.Step` godocs, which mention that 'core steps' are steps which have no sub-steps, which are exactly what that factory builds.

The overall idea was to avoid using the term 'builder' entirely since it can be easily confused with builds and is kind of just another name for `Factory`. I was tempted to rename `Stepper` to `StepFactory` but I really didn't want to end up with `engine.StepFactoryFactory`. 😂 It's just a function alias anyway.

I'll go ahead and merge but lemme know what you think of the new names. Thanks for the reviews!
